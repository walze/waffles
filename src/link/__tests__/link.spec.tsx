import { useRef, useEffect } from 'react';
import { render, fireEvent } from '@testing-library/react';

import { Link } from '../index';
import { Plus, ExternalLink } from '../../icon';

const sizes = ['small', 'medium', 'large', 'inherit'] as const;

function TestRefLink() {
  const linkRef = useRef<HTMLAnchorElement>(null);

  useEffect(() => {
    linkRef.current?.focus();
  }, []);

  return (
    <Link href="https://datacamp.com" ref={linkRef}>
      Focused With Ref
    </Link>
  );
}

describe('Link', () => {
  it('renders a link containing the texts', () => {
    const { container } = render(
      <Link href="https://taylor-swift-fanclub.com">
        Join Taylor Swift Fanclub
      </Link>,
    );

    const link = container.querySelector('a');

    expect(link).toBeInTheDocument();
    expect(link).toHaveAttribute('href', 'https://taylor-swift-fanclub.com');
    expect(link).toHaveTextContent('Join Taylor Swift Fanclub');
  });

  it('sets the data attribute on the link', () => {
    const { getByTestId } = render(
      <Link data-testid="taylors-link">Join Taylor Swift Fanclub</Link>,
    );

    const link = getByTestId('taylors-link');

    expect(link).toBeInTheDocument();
  });

  it('handles focus correctly', () => {
    const handleFocus = jest.fn();
    const { getByText } = render(
      <Link onFocus={handleFocus} href="https://taylor-swift-fanclub.com">
        Join Taylor Swift Fanclub
      </Link>,
    );

    const link = getByText('Join Taylor Swift Fanclub');
    fireEvent.focus(link);

    expect(handleFocus).toHaveBeenCalledTimes(1);
  });

  it('renders left icon', () => {
    const { container, getByText } = render(
      <Link iconLeft={<Plus />}>Go to Fanpage</Link>,
    );

    const link = getByText('Go to Fanpage').closest('a');
    const icon = container.querySelector('svg');

    expect(link).toBeInTheDocument();
    expect(link).toContainElement(icon);
  });

  it('renders right icon', () => {
    const { container, getByText } = render(
      <Link iconRight={<ExternalLink />}>Read More</Link>,
    );

    const link = getByText('Read More').closest('a');
    const icon = container.querySelector('svg');

    expect(link).toBeInTheDocument();
    expect(link).toContainElement(icon);
  });

  it('renders both left and right icons', () => {
    const { container, getByText } = render(
      <Link iconLeft={<Plus />} iconRight={<ExternalLink />}>
        Go to Another Fanpage
      </Link>,
    );

    const link = getByText('Go to Another Fanpage');
    const icons = container.querySelectorAll('svg');

    expect(link).toBeInTheDocument();
    expect(icons).toHaveLength(2);
  });

  it('renders icon of custom size', () => {
    const { container, getByText } = render(
      <Link iconRight={<ExternalLink size="xsmall" />}>
        Go to Another Fanpage
      </Link>,
    );

    const link = getByText('Go to Another Fanpage').closest('a');
    const icon = container.querySelector('svg');

    expect(link).toBeInTheDocument();
    expect(link).toContainElement(icon);
    expect(icon).toHaveAttribute('width', '12');
    expect(icon).toHaveAttribute('height', '12');
  });

  it('renders snapshot of focused state', () => {
    const { getByText } = render(<Link href="https://datacamp.com">Test</Link>);

    const link = getByText('Test').closest('a') as HTMLAnchorElement;
    fireEvent.focus(link);

    expect(link).toMatchSnapshot();
  });

  it('accepts ref and could be focused programmatically when link has a "href" attribute', () => {
    const { container } = render(<TestRefLink />);

    const link = container.querySelector('a');

    expect(link).toHaveFocus();
  });

  describe('renders snapshot with icons of ', () => {
    sizes.forEach((size) => {
      it(`${size} size`, () => {
        const { container } = render(
          <Link size={size} iconLeft={<Plus />} iconRight={<Plus />}>
            Join Ariana Grande Fanclub
          </Link>,
        );

        const link = container.querySelector('a');

        expect(link).toMatchSnapshot();
      });
    });
  });

  describe('renders snapshot of inverted variant', () => {
    const { container } = render(
      <Link inverted href="https://taylor-swift-fanclub.com">
        Join Taylor Swift Fanclub
      </Link>,
    );

    const link = container.querySelector('a');

    expect(link).toMatchSnapshot();
  });
});
