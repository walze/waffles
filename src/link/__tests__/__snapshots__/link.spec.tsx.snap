// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
.emotion-0 {
  display: inline;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-weight: 800;
  outline: 0;
  border-radius: 4px;
  -webkit-box-decoration-break: clone;
  box-decoration-break: clone;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  text-decoration-color: transparent;
  -webkit-transition: box-shadow 125ms ease-out,background-color 125ms ease-out,text-decoration-color 200ms ease-out;
  transition: box-shadow 125ms ease-out,background-color 125ms ease-out,text-decoration-color 200ms ease-out;
  color: #06BDFC;
  font-size: inherit;
  line-height: inherit;
}

.emotion-0 svg {
  vertical-align: middle;
  margin-top: -2px;
}

.emotion-0:hover {
  background-color: rgba(255, 255, 255, 0.1);
}

.emotion-0:hover {
  text-decoration-color: #06BDFC;
  box-shadow: 0 0 0 2px rgba(255, 255, 255, 0.1);
}

<a
  class="emotion-0"
  href="https://taylor-swift-fanclub.com"
>
  <span
    class="emotion-1"
  >
    Join Taylor Swift Fanclub
  </span>
</a>
`;

exports[`Link renders snapshot of focused state 1`] = `
.emotion-0 {
  display: inline;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-weight: 800;
  outline: 0;
  border-radius: 4px;
  -webkit-box-decoration-break: clone;
  box-decoration-break: clone;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  text-decoration-color: transparent;
  -webkit-transition: box-shadow 125ms ease-out,background-color 125ms ease-out,text-decoration-color 200ms ease-out;
  transition: box-shadow 125ms ease-out,background-color 125ms ease-out,text-decoration-color 200ms ease-out;
  color: #0075AD;
  font-size: inherit;
  line-height: inherit;
  box-shadow: 0 0 0 2px #009BD8;
}

.emotion-0 svg {
  vertical-align: middle;
  margin-top: -2px;
}

.emotion-0:hover {
  background-color: rgba(5, 25, 45, 0.1);
}

<a
  class="emotion-0"
  href="https://datacamp.com"
>
  <span
    class="emotion-1"
  >
    Test
  </span>
</a>
`;

exports[`Link renders snapshot with icons of  inherit size 1`] = `
.emotion-0 {
  display: inline;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-weight: 800;
  outline: 0;
  border-radius: 4px;
  -webkit-box-decoration-break: clone;
  box-decoration-break: clone;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  text-decoration-color: transparent;
  -webkit-transition: box-shadow 125ms ease-out,background-color 125ms ease-out,text-decoration-color 200ms ease-out;
  transition: box-shadow 125ms ease-out,background-color 125ms ease-out,text-decoration-color 200ms ease-out;
  color: #0075AD;
  font-size: inherit;
  line-height: inherit;
}

.emotion-0 svg {
  vertical-align: middle;
  margin-top: -2px;
}

.emotion-0:hover {
  background-color: rgba(5, 25, 45, 0.1);
}

.emotion-0:hover {
  text-decoration-color: #0075AD;
  box-shadow: 0 0 0 2px rgba(5, 25, 45, 0.1);
}

.emotion-1 {
  padding-left: 4px;
  padding-right: 4px;
}

<a
  class="emotion-0"
>
  <svg
    aria-hidden="true"
    height="16"
    viewBox="0 0 18 18"
    width="16"
  >
    <path
      clip-rule="evenodd"
      d="M9 3c-.6 0-1 .5-1 1.2V8H4.2C3.5 8 3 8.4 3 9s.5 1 1.2 1H8v3.8c0 .7.4 1.2 1 1.2s1-.5 1-1.2V10h3.8c.7 0 1.2-.4 1.2-1s-.5-1-1.2-1H10V4.2C10 3.5 9.6 3 9 3z"
      fill="currentColor"
      fill-rule="evenodd"
    />
  </svg>
  <span
    class="emotion-1"
  >
    Join Ariana Grande Fanclub
  </span>
  <svg
    aria-hidden="true"
    height="16"
    viewBox="0 0 18 18"
    width="16"
  >
    <path
      clip-rule="evenodd"
      d="M9 3c-.6 0-1 .5-1 1.2V8H4.2C3.5 8 3 8.4 3 9s.5 1 1.2 1H8v3.8c0 .7.4 1.2 1 1.2s1-.5 1-1.2V10h3.8c.7 0 1.2-.4 1.2-1s-.5-1-1.2-1H10V4.2C10 3.5 9.6 3 9 3z"
      fill="currentColor"
      fill-rule="evenodd"
    />
  </svg>
</a>
`;

exports[`Link renders snapshot with icons of  large size 1`] = `
.emotion-0 {
  display: inline;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-weight: 800;
  outline: 0;
  border-radius: 4px;
  -webkit-box-decoration-break: clone;
  box-decoration-break: clone;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  text-decoration-color: transparent;
  -webkit-transition: box-shadow 125ms ease-out,background-color 125ms ease-out,text-decoration-color 200ms ease-out;
  transition: box-shadow 125ms ease-out,background-color 125ms ease-out,text-decoration-color 200ms ease-out;
  color: #0075AD;
  font-size: 16px;
  line-height: 1.5;
}

.emotion-0 svg {
  vertical-align: middle;
  margin-top: -2px;
}

.emotion-0:hover {
  background-color: rgba(5, 25, 45, 0.1);
}

.emotion-0:hover {
  text-decoration-color: #0075AD;
  box-shadow: 0 0 0 2px rgba(5, 25, 45, 0.1);
}

.emotion-1 {
  padding-left: 4px;
  padding-right: 4px;
}

<a
  class="emotion-0"
>
  <svg
    aria-hidden="true"
    height="16"
    viewBox="0 0 18 18"
    width="16"
  >
    <path
      clip-rule="evenodd"
      d="M9 3c-.6 0-1 .5-1 1.2V8H4.2C3.5 8 3 8.4 3 9s.5 1 1.2 1H8v3.8c0 .7.4 1.2 1 1.2s1-.5 1-1.2V10h3.8c.7 0 1.2-.4 1.2-1s-.5-1-1.2-1H10V4.2C10 3.5 9.6 3 9 3z"
      fill="currentColor"
      fill-rule="evenodd"
    />
  </svg>
  <span
    class="emotion-1"
  >
    Join Ariana Grande Fanclub
  </span>
  <svg
    aria-hidden="true"
    height="16"
    viewBox="0 0 18 18"
    width="16"
  >
    <path
      clip-rule="evenodd"
      d="M9 3c-.6 0-1 .5-1 1.2V8H4.2C3.5 8 3 8.4 3 9s.5 1 1.2 1H8v3.8c0 .7.4 1.2 1 1.2s1-.5 1-1.2V10h3.8c.7 0 1.2-.4 1.2-1s-.5-1-1.2-1H10V4.2C10 3.5 9.6 3 9 3z"
      fill="currentColor"
      fill-rule="evenodd"
    />
  </svg>
</a>
`;

exports[`Link renders snapshot with icons of  medium size 1`] = `
.emotion-0 {
  display: inline;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-weight: 800;
  outline: 0;
  border-radius: 4px;
  -webkit-box-decoration-break: clone;
  box-decoration-break: clone;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  text-decoration-color: transparent;
  -webkit-transition: box-shadow 125ms ease-out,background-color 125ms ease-out,text-decoration-color 200ms ease-out;
  transition: box-shadow 125ms ease-out,background-color 125ms ease-out,text-decoration-color 200ms ease-out;
  color: #0075AD;
  font-size: 14px;
  line-height: 1.5;
}

.emotion-0 svg {
  vertical-align: middle;
  margin-top: -2px;
}

.emotion-0:hover {
  background-color: rgba(5, 25, 45, 0.1);
}

.emotion-0:hover {
  text-decoration-color: #0075AD;
  box-shadow: 0 0 0 2px rgba(5, 25, 45, 0.1);
}

.emotion-1 {
  padding-left: 4px;
  padding-right: 4px;
}

<a
  class="emotion-0"
>
  <svg
    aria-hidden="true"
    height="16"
    viewBox="0 0 18 18"
    width="16"
  >
    <path
      clip-rule="evenodd"
      d="M9 3c-.6 0-1 .5-1 1.2V8H4.2C3.5 8 3 8.4 3 9s.5 1 1.2 1H8v3.8c0 .7.4 1.2 1 1.2s1-.5 1-1.2V10h3.8c.7 0 1.2-.4 1.2-1s-.5-1-1.2-1H10V4.2C10 3.5 9.6 3 9 3z"
      fill="currentColor"
      fill-rule="evenodd"
    />
  </svg>
  <span
    class="emotion-1"
  >
    Join Ariana Grande Fanclub
  </span>
  <svg
    aria-hidden="true"
    height="16"
    viewBox="0 0 18 18"
    width="16"
  >
    <path
      clip-rule="evenodd"
      d="M9 3c-.6 0-1 .5-1 1.2V8H4.2C3.5 8 3 8.4 3 9s.5 1 1.2 1H8v3.8c0 .7.4 1.2 1 1.2s1-.5 1-1.2V10h3.8c.7 0 1.2-.4 1.2-1s-.5-1-1.2-1H10V4.2C10 3.5 9.6 3 9 3z"
      fill="currentColor"
      fill-rule="evenodd"
    />
  </svg>
</a>
`;

exports[`Link renders snapshot with icons of  small size 1`] = `
.emotion-0 {
  display: inline;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-weight: 800;
  outline: 0;
  border-radius: 4px;
  -webkit-box-decoration-break: clone;
  box-decoration-break: clone;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  text-decoration-color: transparent;
  -webkit-transition: box-shadow 125ms ease-out,background-color 125ms ease-out,text-decoration-color 200ms ease-out;
  transition: box-shadow 125ms ease-out,background-color 125ms ease-out,text-decoration-color 200ms ease-out;
  color: #0075AD;
  font-size: 12px;
  line-height: 1.5;
}

.emotion-0 svg {
  vertical-align: middle;
  margin-top: -2px;
}

.emotion-0:hover {
  background-color: rgba(5, 25, 45, 0.1);
}

.emotion-0:hover {
  text-decoration-color: #0075AD;
  box-shadow: 0 0 0 2px rgba(5, 25, 45, 0.1);
}

.emotion-1 {
  padding-left: 2px;
  padding-right: 2px;
}

<a
  class="emotion-0"
>
  <svg
    aria-hidden="true"
    height="14"
    viewBox="0 0 18 18"
    width="14"
  >
    <path
      clip-rule="evenodd"
      d="M9 3c-.6 0-1 .5-1 1.2V8H4.2C3.5 8 3 8.4 3 9s.5 1 1.2 1H8v3.8c0 .7.4 1.2 1 1.2s1-.5 1-1.2V10h3.8c.7 0 1.2-.4 1.2-1s-.5-1-1.2-1H10V4.2C10 3.5 9.6 3 9 3z"
      fill="currentColor"
      fill-rule="evenodd"
    />
  </svg>
  <span
    class="emotion-1"
  >
    Join Ariana Grande Fanclub
  </span>
  <svg
    aria-hidden="true"
    height="14"
    viewBox="0 0 18 18"
    width="14"
  >
    <path
      clip-rule="evenodd"
      d="M9 3c-.6 0-1 .5-1 1.2V8H4.2C3.5 8 3 8.4 3 9s.5 1 1.2 1H8v3.8c0 .7.4 1.2 1 1.2s1-.5 1-1.2V10h3.8c.7 0 1.2-.4 1.2-1s-.5-1-1.2-1H10V4.2C10 3.5 9.6 3 9 3z"
      fill="currentColor"
      fill-rule="evenodd"
    />
  </svg>
</a>
`;
